### PR TITLE
Fix compression of punctuation marks at line start of end

### LINKED ISSUE #3743: Compression of punctuation marks at line end is not work in `simple` linebreak mode
### Description

Acording to **Compression of punctuation marks at line start or line end** in [W3C:Requirements for Chinese Text Layout](https://www.w3.org/TR/clreq/#punctuation_width_adjustment)

> According to section 5.1.10 of the General Rules for Punctuation (GB/T 15834—2011), when a full-width punctuation character appears at the end of a line, it SHALL be trimmed to half-width.

But the punctuation at line end is not compressed when in `simple` linebreak mode.

截屏2024-03-21 11 48 17

The last `。` need to be compressed in the first line.

### Reproduction URL

https://typst.app/project/rRsNaXqWFHfMXRYhTEiybG

### Operating system

macOS

### Typst version

- [X] I am using the latest version of Typst

### PR DESCRIPTION
**Pull request by Jocs**
fix #3743 No need to add `stretchability` of the first or last CJK punctuation glyph to avoid stretching in line adjustment.

### CODE REVIEW

**Comment by laurmaedje** [comment]
cc: @peng1999 who is most familiar with this part of the codebase

File: crates/typst/src/layout/inline/mod.rs [hunk 1]
```diff
@@ -1085,7 +1085,9 @@ fn line<'a>(
                         // Compression of punctuation marks at line start or line end
                         let shrink_amount = last_glyph.shrinkability().1;
                         let punct = reshaped.glyphs.to_mut().last_mut().unwrap();
-                        punct.shrink_right(shrink_amount);
+                        // No need to add `stretchability` of the last CJK punctuation glyph to avoid stretching in line adjustment.
+                        punct.x_advance -= shrink_amount;
+                        punct.adjustability.shrinkability.1 -= shrink_amount;
```
Comment (peng1999) [inline crates/typst/src/layout/inline/mod.rs:1090]: There are three places of adjustments:

- At line start;
- At line end;
- Between two consecutive punctuation marks.

All of the three places shouldn't add the stretchability. So maybe you can just change the implementation of `shrink_right` and `shrink_left`.

Comment (Jocs) [reply crates/typst/src/layout/inline/mod.rs:1090]: Ok, I'll check it, and update this PR, Thanks.

Comment (Jocs) [reply crates/typst/src/layout/inline/mod.rs]: @peng1999 I have modified `shrink_right` and `shrink_left`, and it works.

**Comment by Jocs** [comment]
@peng1999 Please review it again.

**Comment by peng1999** [comment]
I think its better to add a test for the punctuation adjustment behavior when line is underfull.

You can construct a line with following property:

- Inter-character space expansion
- Have punctuation marks on line start and end
- Have consecutive punctuation marks within

image

The above is my quick sketch from Microsoft Word.

File: tests/typ/layout/par-justify-cjk.typ [hunk 2]
```diff
@@ -49,3 +49,16 @@
...
+
+// The test case includes the following scenarios:
+// - Compression of punctuation marks at line start or line end
+// - Adjustment of adjacent punctuation marks
+
+#set page(width: 110pt + 10pt, margin: (x: 5pt))
+#set text(lang: "zh", font: "Noto Serif CJK SC")
+#set par(justify: true)
+
+标注在字间的标点符号（乙式括号省略号以外）通常占一个汉字宽度，使其易于识别、适合配置及排版，有些排版风格完全不对标点宽度进行任何调整。但是为了让文字体裁更加紧凑易读，，，以及执行3.1.4 行首行尾禁则时，就需要对标点符号的宽度进行调整。是否调整取决于对排版风格的判断，台湾的很多印刷品都采用不调整的风格；而在中国大陆和香港的出版物中，多数采取调整的风格。标点符号宽度调整通常分为两种情形：1. 标点符号连续出现 2.标点出现在行首或行尾。调整时的风格不尽相同，下文仅阐述基本原则。
```
Comment (peng1999) [inline tests/typ/layout/par-justify-cjk.typ:64]: Text after “是否调整取决于……” can be removed as it has the same behavior in Typst 0.11.0.

Comment (Jocs) [reply tests/typ/layout/par-justify-cjk.typ]: done

**Comment by peng1999** [comment]
LGTM

**Comment by laurmaedje** [comment]
Thanks to both of you!

**Pull request merged**